pkg/scheduler/dependency_graph.go

```diff
@@ -434,37 +482,74 @@ func (sched *Scheduler) BuildDependencyGraph(options interfaces.DependencyGraphO
 	depGraph := NewDependencyGraph(sched, options)
 	rootContext := &GraphContext{scheduler: sched, graph: depGraph, flow: flow, args: options.Args}
 
-	if _, ok := dependencies[fullFlowName]; !ok {
-		log.Printf("Flow %s is empty", options.FlowName)
-		return depGraph, nil
+	replicas, deleteReplicas, err := sched.allocateReplicas(flow, options)
+	if err != nil {
+		return nil, err
 	}
 
-	type Block struct {
-		dependency        client.Dependency
-		scheduledResource *ScheduledResource
-		parentContext     *GraphContext
+	err = sched.fillDependencyGraph(rootContext, resDefs, dependencies, flow, replicas, false)
+	if err != nil {
+		return nil, err
 	}
-
-	replicas, err := sched.allocateReplicas(flow, options)
+	err = sched.fillDependencyGraph(rootContext, resDefs, dependencies, flow, deleteReplicas, true)
 	if err != nil {
 		return nil, err
 	}
 
+	for _, value := range depGraph.graph {
+		value.RequiredBy = unique(value.RequiredBy)
+		value.Requires = unique(value.Requires)
+		value.usedInReplicas = unique(value.usedInReplicas)
+	}
+
+	if len(deleteReplicas) > 0 {
```

Comment:
what is `getCorrectFlow`? How function that creates dependency graph can be a method of DependencyGraph? Could you please elaborate on your idea?
Splitting things is not always a good idea. The cost of split is that now you have to pass a lot of state back and forth, create new data structures etc. Sometimes, when you try to extract parts of code into separate functions you get function that has no clear purpose, it is almost impossible to find good name for such functions. Sometimes after such refactoring, despite having smaller functions it becomes harder to understand the big picture. It might very well be the case here. The block of code here is just 4 lines of code. Creating another abstraction for this may introduce more complexity than it removes.
Also, it might be a better idea to do such refactoring in a separate commit in order to minimize number of changes here